File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/Fnv1a32.java
package com.freight.hash;

/** fnv1a32 over raw bytes. */
public final class Fnv1a32 implements HashAlgorithm {

    @Override
    public String name() {
        return "fnv1a32";
    }

    @Override
    public long apply(byte[] data) {
        long state = 2166136261L;
        for (int i = 0; i < data.length; i++) {
            state ^= (data[i] & 0xFF);
            state = (state * 16777619L) & 0xFFFFFFFFL;
        }
        return state;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/fnv1a32.cpp
#include "freight/hashes.h"

namespace freight {

// fnv1a32 over raw bytes.
uint64_t hash_fnv1a32(const std::string& data) {
  uint32_t state = 2166136261u;
  for (size_t i = 0; i < data.size(); ++i) {
    state ^= static_cast<uint8_t>(data[i]);
    state *= 16777619u;
  }
  return static_cast<uint64_t>(state);
}

}  // namespace freight
